### PR TITLE
Nolio fork pull request

### PR DESCRIPTION
**Pull request by Thomas-Plazas**

### CODE REVIEW

File: src/Cloud/OAuthDialog.cpp [hunk 1]
```diff
@@ -29,6 +29,7 @@
 #include "PolarFlow.h"
 
 #include <QJsonParseError>
+#include <QDebug>
```
Comment (liversedge) [inline src/Cloud/OAuthDialog.cpp:32]: please remove

File: src/Cloud/OAuthDialog.cpp [hunk 2]
```diff
@@ -102,15 +104,22 @@ OAuthDialog::OAuthDialog(Context *context, OAuthSite site, CloudService *service
...
+
+        urlstr = QString("https://www.nolio.io/api/authorize/?");
+        urlstr.append("client_id=").append(GC_NOLIO_CLIENT_ID).append("&");
+        urlstr.append("redirect_uri=http://www.goldencheetah.org/&");
+        urlstr.append("response_type=code");
+
     } else if (site == DROPBOX) {
 
-        urlstr = QString("https://www.dropbox.com/oauth2/authorize?");
+            urlstr = QString("https://www.dropbox.com/oauth2/authorize?");
```
Comment (liversedge) [inline src/Cloud/OAuthDialog.cpp:116]: there is new whitespace in here and the next few lines to remove.

File: src/Cloud/OAuthDialog.cpp [hunk 3]
```diff
@@ -435,7 +451,7 @@ OAuthDialog::networkRequestFinished(QNetworkReply *reply)
 
     // we've been told to ignore responses (used by POLAR, maybe others in future)
     if (ignore) return;
-
+    bool is_nolio = site == NOLIO;
```
Comment (liversedge) [inline src/Cloud/OAuthDialog.cpp:454]: is_nolio does not appear to be used and isn't adding much value anyway, please remove

File: src/Cloud/OAuthDialog.cpp [hunk 4]
```diff
@@ -610,10 +625,17 @@ OAuthDialog::networkRequestFinished(QNetworkReply *reply)
...
+            appsettings->setValue(GC_NOLIO_REFRESH_TOKEN, refresh_token);
+            appsettings->setValue(GC_NOLIO_LAST_REFRESH, QDateTime::currentDateTime());
+            QString info = QString(tr("Nolio authorization was successful."));
+            QMessageBox information(QMessageBox::Information, tr("Information"), info);
+            information.exec();
         }
 
     } else {
-
+            qDebug() << "error  111";
```
Comment (liversedge) [inline src/Cloud/OAuthDialog.cpp:638]: developer debug should be removed

File: src/src.pro [hunk 5]
```diff
@@ -19,8 +19,10 @@ greaterThan(QT_MAJOR_VERSION, 5) {
 include(gcconfig.pri)
 
 # You can also define your own local source to add to build
-HEADERS += $${LOCALHEADERS}
-SOURCES += $${LOCALSOURCES}
+HEADERS += $${LOCALHEADERS} \
+    Cloud/Nolio.h
+SOURCES += $${LOCALSOURCES} \
+    Cloud/Nolio.cpp
```
Comment (liversedge) [inline src/src.pro:26]: these should now be added to the proper lines in src.pro (around L615)

**Comment by liversedge** [comment]
I have added some basic code comments, but have not tested the change.
The commit requires rebasing to squash and merge commits (see: git rebase --i) 

@grauser is testing this and will likely feedback if there are any defects etc

**Comment by liversedge** [comment]
Please also remove the .idea folder from the commit.

File: src/Cloud/Nolio.cpp [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+    if (parseError.error == QJsonParseError::NoError) {
+        QJsonObject activity = document.object();
+        QDateTime start_time = QDateTime::fromString(activity["date_start"].toString(), Qt::ISODate);
+        RideFile *ride = new RideFile(start_time.toUTC(), 1.0f);
+        // set nolio id in metadata
+        if (!activity["nolio_id"].isNull()) ride->setTag("NolioID",  QString("%1").arg(activity["id"].toVariant().toULongLong()));
+        // set sport and subsport
+        if (!activity["sport_id"].isNull()) {
+            double sport_id = activity["sport_id"].toDouble();
+            if (sport_id == 14 or sport_id == 15 or sport_id == 18 or sport_id == 28 or sport_id == 35 or sport_id == 36) ride->setTag("Sport", "Bike");
```
Comment (liversedge) [inline src/Cloud/Nolio.cpp:262]: or is not a valid keyword in C++ use || this breaks on windows builds, it is repeated for multiple lines.

**Review by grauser — CHANGES_REQUESTED** [review]
I tested the PR and made some suggestions in Nolio.cpp and Nolio.h

**Review by grauser — APPROVED** [review]

**Comment by liversedge** [comment]
Thanks Thomas, and thanks Damien for the review!

**Pull request merged**